### ISSUE TITLE
Memory allocation error with Caret Wrapper

### ISSUE THREAD

**Issue by larry77** [dmlc/xgboost#290]
The following script causes the a memory allocation error when you let it run long enough.
The train. csv data can be downloaded from the otto classification

 http://bit.ly/1I2bauw

```r
library(readr)
library(xgboost)
library(caret)

modelInfo <- list(label = "eXtreme Gradient Boosting",
                  library = c("xgboost", "plyr"),
                  type = c("Regression", "Classification"),
                  parameters = data.frame(parameter = c('nrounds', 'max_depth', 'eta'),
                                          class = rep("numeric", 3),
                                          label = c('# Boosting Iterations', 'Max Tree Depth', 
                                                    'Shrinkage')),
                  grid = function(x, y, len = NULL) expand.grid(max_depth = seq(1, len),
                                                                nrounds = floor((1:len) * 50),
                                                                eta = .3),
                  loop = function(grid) {     
                    loop <- ddply(grid, c("eta", "max_depth"),
                                  function(x) c(nrounds = max(x$nrounds)))
                    submodels <- vector(mode = "list", length = nrow(loop))
                    for(i in seq(along = loop$nrounds)) {
                      index <- which(grid$max_depth == loop$max_depth[i] & 
                                       grid$eta == loop$eta[i])
                      trees <- grid[index, "nrounds"] 
                      submodels[[i]] <- data.frame(nrounds = trees[trees != loop$nrounds[i]])
                    }    
                    list(loop = loop, submodels = submodels)
                  },
                  fit = function(x, y, wts, param, lev, last, classProbs, ...) { 
                    if(is.factor(y)) {
                      if(length(lev) == 2) {
                        y <- ifelse(y == lev[1], 1, 0) 
                        dat <- xgb.DMatrix(as.matrix(x), label = y)
                        out <- xgb.train(list(eta = param$eta, 
                                              max_depth = param$max_depth), 
                                         data = dat,
                                         nrounds = param$nrounds,
                                         objective = "binary:logistic",
                                         ...)
                      } else {
                        y <- as.numeric(y) - 1
                        dat <- xgb.DMatrix(as.matrix(x), label = y)
                        out <- xgb.train(list(eta = param$eta, 
                                              max_depth = param$max_depth), 
                                         data = dat,
                                         num_class = length(lev),
                                         nrounds = param$nrounds,
                                         objective = "multi:softprob",
                                         ...)
                      }     
                    } else {
                      dat <- xgb.DMatrix(as.matrix(x), label = y)
                      out <- xgb.train(list(eta = param$eta, 
                                            max_depth = param$max_depth), 
                                       data = dat,
                                       nrounds = param$nrounds,
                                       objective = "reg:linear",
                                       ...)
                    }
                    out
                  },
   predict = function(modelFit, newdata, submodels = NULL) {
                    newdata <- xgb.DMatrix(as.matrix(newdata))
                    out <- predict(modelFit, newdata)
                    if(modelFit$problemType == "Classification") {
                      if(length(modelFit$obsLevels) == 2) {
                        out <- ifelse(out >= .5, 
                                      modelFit$obsLevels[1], 
                                      modelFit$obsLevels[2])
                      } else {
                        out <- matrix(out, ncol = length(modelFit$obsLevels), byrow = TRUE)
                        out <- modelFit$obsLevels[apply(out, 1, which.max)]
                      }
                    }
  if(!is.null(submodels)) {
                      tmp <- vector(mode = "list", length = nrow(submodels) + 1)
                      tmp[[1]] <- out
                      for(j in seq(along = submodels$nrounds)) {
                        tmp_pred <- predict(modelFit, newdata, ntreelimit = submodels$nrounds[j])
                        if(modelFit$problemType == "Classification") {
                          if(length(modelFit$obsLevels) == 2) {
                            tmp_pred <- ifelse(tmp_pred >= .5, 
                                               modelFit$obsLevels[1], 
                                               modelFit$obsLevels[2])
                          } else {
                            tmp_pred <- matrix(tmp_pred, ncol = length(modelFit$obsLevels), byrow = TRUE)
                            tmp_pred <- modelFit$obsLevels[apply(tmp_pred, 1, which.max)]
                          }
                        }
                        tmp[[j+1]]  <- tmp_pred
                      }
                      out <- tmp
                    }
                    out  
                  },
                  prob = function(modelFit, newdata, submodels = NULL) {
                    newdata <- xgb.DMatrix(as.matrix(newdata))
                    out <- predict(modelFit, newdata)
                    if(length(modelFit$obsLevels) == 2) {
                      out <- cbind(out, 1 - out)
                      colnames(out) <- modelFit$obsLevels
                    } else {
                      out <- matrix(out, ncol = length(modelFit$obsLevels), byrow = TRUE)
                      colnames(out) <- modelFit$obsLevels
                    }
                    out <- as.data.frame(out)
                    if(!is.null(submodels)) {
                      tmp <- vector(mode = "list", length = nrow(submodels) + 1)
                      tmp[[1]] <- out
                      for(j in seq(along = submodels$nrounds)) {
                        tmp_pred <- predict(modelFit, newdata, ntreelimit = submodels$nrounds[j])
                        if(length(modelFit$obsLevels) == 2) {
                          tmp_pred <- cbind(tmp_pred, 1 - tmp_pred)
                          colnames(tmp_pred) <- modelFit$obsLevels
                        } else {
                          tmp_pred <- matrix(tmp_pred, ncol = length(modelFit$obsLevels), byrow = TRUE)
                          colnames(tmp_pred) <- modelFit$obsLevels
                        }
                        tmp_pred <- as.data.frame(tmp_pred)
                        tmp[[j+1]]  <- tmp_pred
                      }
                      out <- tmp
                    }
                    out  
                  },
                  predictors = function(x, ...) {
                    imp <- xgb.importance(x$xNames, model = x)
                    x$xNames[x$xNames %in% imp$Feature]
                  },
                  varImp = function(object, numTrees = NULL, ...) {
                    imp <- xgb.importance(x$xNames, model = x)
                    imp <- as.data.frame(imp)[, 1:2]
                    rownames(imp) <- as.character(imp[,1])
                    imp <- imp[,2,drop = FALSE]
                    colnames(imp) <- "Overall"
                    imp   
                  },
                  levels = function(x) x$obsLevels,
                  tags = c("Tree-Based Model", "Boosting", "Ensemble Model", "Implicit Feature Selection"),
                  sort = function(x) {
                    # This is a toss-up, but the # trees probably adds
                    # complexity faster than number of splits
                    x[order(x$nrounds, x$max_depth, x$eta),] 
                  })


train <- read_csv("train.csv")

train$target <- as.factor(train$target)
grid <- expand.grid(nrounds = seq(26, 406, by = 20),
                    max_depth = 5:17,
                    eta = c(seq(0.01,0.31,by=0.05)))
mod1 <- train(target~.,data=train,
              method = modelInfo,
              tuneGrid = grid,  nthread=10,
              verbose=T,
              trControl = trainControl(method = "repeatedcv" , number=10,
                  repeats=10,
                  classProbs = TRUE, verbose=T)  )



saveRDS(mod1, "caret-xgboost-otto.RDS")
```

**Comment by tqchen** [comment]
Can you try to provide an simplified case to reproduce the error faster?  
Currently I cannot tell the cause of error and seems I could not reproduce it on my machine.
 For example, try to reduce numrounds in the tuning parameter

Thanks

**Comment by larry77** [comment]
On Thu, May 07, 2015 at 10:39:09PM -0700, Tianqi Chen wrote:
>Can you try to provide an simplified case to reproduce the error faster?
>Currently I cannot tell the cause of error and seems I could not reproduce it on my machine.
> For example, try to reduce numrounds in the tuning parameter
>
>Thanks
>

Hi,
I wish it was that easy. When the model is small (i.e. not many grid
values and not many iterations), then it completes successfully (I
paste at the end of the email an example which runs without a glitch
and should take 3-4
minutes on a laptop). However, in the full scale simulation
that I emailed you, it breaks down after some time on a workstation
with a segmentation fault error message.
I have not yet been able to reproduce the error with a small toy
model.

Lorenzo

###############################################################################################


library(readr)
library(xgboost)
library(caret)

modelInfo <- list(label = "eXtreme Gradient Boosting",
                  library = c("xgboost", "plyr"),
		                    type = c("Regression",
                  "Classification"),
		                    parameters = data.frame(parameter
                  = c('nrounds', 'max_depth', 'eta'),
		                                            class =
                  rep("numeric", 3),
		                                            label =
                  c('# Boosting Iterations', 'Max Tree Depth',
		                                                      'Shrinkage')),
								                        grid
                  = function(x, y, len = NULL) expand.grid(max_depth =
                  seq(1, len),
		                                                                  nrounds
                  = floor((1:len) * 50),
		                                                                  eta
                  = .3),
		                    loop = function(grid) {
				                        loop <-
                  ddply(grid, c("eta", "max_depth"),
		                                    function(x)
                  c(nrounds = max(x$nrounds)))
		                      submodels <- vector(mode =
                  "list", length = nrow(loop))
		                      for(i in seq(along =
                  loop$nrounds)) {
		                        index <- which(grid$max_depth
                  == loop$max_depth[i] &
		                                         grid$eta ==
                  loop$eta[i])
		                        trees <- grid[index,
                  "nrounds"]
		                        submodels[[i]] <-
                  data.frame(nrounds = trees[trees !=
                  loop$nrounds[i]])
		                      }
				                          list(loop =
                  loop, submodels = submodels)
		                    },
				                      fit =
                  function(x, y, wts, param, lev, last, classProbs,
                  ...) {
		                      if(is.factor(y)) {
				                            if(length(lev)
                  == 2) {
		                          y <- ifelse(y == lev[1], 1,
                  0)
		                          dat <-
                  xgb.DMatrix(as.matrix(x), label = y)
		                          out <- xgb.train(list(eta =
                  param$eta,
		                                                max_depth
                  = param$max_depth),
		                                           data = dat,
							                                            nrounds
                  = param$nrounds,
		                                           objective =
                  "binary:logistic",
		                                           ...)
							                         }
                  else {
		                          y <- as.numeric(y) - 1
					                          dat
                  <- xgb.DMatrix(as.matrix(x), label = y)
		                          out <- xgb.train(list(eta =
                  param$eta,
		                                                max_depth
                  = param$max_depth),
		                                           data = dat,
							                                            num_class
                  = length(lev),
		                                           nrounds =
                  param$nrounds,
		                                           objective =
                  "multi:softprob",
		                                           ...)
							                         }
										                     }
                  else {
		                        dat <-
                  xgb.DMatrix(as.matrix(x), label = y)
		                        out <- xgb.train(list(eta =
                  param$eta,
		                                              max_depth
                  = param$max_depth),
		                                         data = dat,
							                                        nrounds
                  = param$nrounds,
		                                         objective =
                  "reg:linear",
		                                         ...)
							                     }
									                         out


                  },
		                    predict = function(modelFit,
		                    newdata, submodels = NULL) {
				                        newdata <-
		                    xgb.DMatrix(as.matrix(newdata))
				                        out <-
		                    predict(modelFit, newdata)
				                        if(modelFit$problemType
		                    == "Classification") {
				                          if(length(modelFit$obsLevels)
		                    == 2) {
				                            out <-
		                    ifelse(out >= .5,
				                                          modelFit$obsLevels[1],
									                                        modelFit$obsLevels[2])
														                      }
		                    else {
				                            out <-
		                    matrix(out, ncol =
		                    length(modelFit$obsLevels), byrow
		                    = TRUE)
				                            out <-
		                    modelFit$obsLevels[apply(out, 1,
		                    which.max)]
				                          }
							                      }

                    if(!is.null(submodels)) {
		                          tmp <- vector(mode = "list",
		                          length = nrow(submodels) +
		                          1)
					                        tmp[[1]]
		                          <- out
					                        for(j
		                          in seq(along =
		                          submodels$nrounds)) {
					                          tmp_pred
		                          <- predict(modelFit,
		                          newdata, ntreelimit =
		                          submodels$nrounds[j])
					                          if(modelFit$problemType
		                          == "Classification") {
					                            if(length(modelFit$obsLevels)
		                          == 2) {
					                              tmp_pred
		                          <- ifelse(tmp_pred >= .5,
					                                                 modelFit$obsLevels[1],
											                                                modelFit$obsLevels[2])
																	                          }
		                          else {
					                              tmp_pred
		                          <- matrix(tmp_pred, ncol =
		                          length(modelFit$obsLevels),
		                          byrow = TRUE)
					                              tmp_pred
		                          <-
		                          modelFit$obsLevels[apply(tmp_pred,
		                          1, which.max)]
					                            }
								                            }
											                            tmp[[j+1]]
		                          <- tmp_pred
					                        }
								                      out
		                          <- tmp
					                      }
							                          out
										                    },
												                      prob
		                          = function(modelFit,
		                          newdata, submodels = NULL) {
					                      newdata
		                          <-
		                          xgb.DMatrix(as.matrix(newdata))
					                      out <-
		                          predict(modelFit, newdata)
					                      if(length(modelFit$obsLevels)
		                          == 2) {
					                        out <-
		                          cbind(out, 1 - out)
					                        colnames(out)
		                          <- modelFit$obsLevels
					                      } else {
							                            out
		                          <- matrix(out, ncol =
		                          length(modelFit$obsLevels),
		                          byrow = TRUE)
					                        colnames(out)
		                          <- modelFit$obsLevels
					                      }
							                          out
		                          <- as.data.frame(out)

                    if(!is.null(submodels)) {
		                          tmp <- vector(mode = "list",
		                          length = nrow(submodels) +
		                          1)
					                        tmp[[1]]
		                          <- out
					                        for(j
		                          in seq(along =
		                          submodels$nrounds)) {
					                          tmp_pred
		                          <- predict(modelFit,
		                          newdata, ntreelimit =
		                          submodels$nrounds[j])
					                          if(length(modelFit$obsLevels)
		                          == 2) {
					                            tmp_pred
		                          <- cbind(tmp_pred, 1 -
		                          tmp_pred)
					                            colnames(tmp_pred)
		                          <- modelFit$obsLevels
					                          }
		                          else {
					                            tmp_pred
		                          <- matrix(tmp_pred, ncol =
		                          length(modelFit$obsLevels),
		                          byrow = TRUE)
					                            colnames(tmp_pred)
		                          <- modelFit$obsLevels
					                          }
								                          tmp_pred
		                          <- as.data.frame(tmp_pred)
					                          tmp[[j+1]]
		                          <- tmp_pred
					                        }
								                      out
		                          <- tmp
					                      }
							                          out
										                    },
												                      predictors
		                          = function(x, ...) {
					                      imp <-
		                          xgb.importance(x$xNames,
		                          model = x)
					                      x$xNames[x$xNames
		                          %in% imp$Feature]
					                    },
							                      varImp
		                          = function(object, numTrees
		                          = NULL, ...) {
					                      imp <-
		                          xgb.importance(x$xNames,
		                          model = x)
					                      imp <-
		                          as.data.frame(imp)[, 1:2]
					                      rownames(imp)
		                          <- as.character(imp[,1])
					                      imp <-
		                          imp[,2,drop = FALSE]
					                      colnames(imp)
		                          <- "Overall"
					                      imp
							                        },
										                  levels
		                          = function(x) x$obsLevels,
					                    tags =
		                          c("Tree-Based Model",
		                          "Boosting", "Ensemble
		                          Model", "Implicit Feature
		                          Selection"),
					                    sort =
		                          function(x) {
					                      # This
		                          is a toss-up, but the #
		                          trees probably adds
					                      #
		                          complexity faster than
		                          number of splits
					                      x[order(x$nrounds,
		                          x$max_depth, x$eta),]
					                    })


#####################################################
#####################################################
#####################################################


## param <- list("objective" = "multi:softprob",
##               "max_depth"=12,
##               "eta"=0.1,
##               "subsample"=0.7,
##               "colsample_bytree"= 1,
##               "gamma"=2,
##                "min_child_weight"=4,
##               "eval_metric" = "mlogloss",
##                "min_child_weight"=2,
##               ## "silent"=1,
##               "num_class" = 9,
##               "nthread" = threads)






train <- read_csv("train.csv")

train$target <- as.factor(train$target)

## x <- subset(train, select=-c(target))

## x <- as.matrix(x)
## x = matrix(as.numeric(x),nrow(x),ncol(x)) ## now x is a matrix of
## purely numerical values


## y <- train$target
## y <- as.factor(y)

## y = gsub('Class_','',y)
## y = as.integer(y)-1 #xgboost take features in [0,numOfClass)
## In other words: y is now a numeric vector in 0...(numOfClass-1)


grid <- expand.grid(nrounds = seq(1, 11, by = 10),
                    max_depth = seq(5,15,by=5),
		                        eta =
                    c(seq(0.01,0.31,by=0.1)))


## grid <- expand.grid(nrounds = c(10),
##                     max_depth = 5:6,
##                     eta = c(seq(0.01,0.06,by=0.05)))


mod1 <- train(target~.,data=train,
              method = modelInfo,
	                    tuneGrid = grid,  nthread=2,
			                  verbose=T,
					                trControl =
              trainControl(method = "repeatedcv" , number=2,
	                        repeats=2,
				                  classProbs = TRUE,
              verbose=T)  )



saveRDS(mod1, "caret-xgboost-otto.RDS")

**Comment by tqchen** [comment]
@larry77 Recently I pushed a bugfix that might fixes some memory issues. Can you try if the newest version in master solves your problem? Thanks

**Comment by tqchen** [comment]
@larry77  Do you have any update on this issue?

**Comment by larry77** [comment]
Hi!
I will let you know shortly.

**Comment by larry77** [comment]
Hi,
The simulation will take, in the best case scenario, a few days to
complete.
What I can say right now is that with the new xgboost it has not
crashed yet, hence for sure something has changed for the better.
I'll post again when the simulation ends or crashes.

**Comment by tqchen** [comment]
great, thanks

**Comment by tqchen** [comment]
@larry77 any updates on this? I think I will close the issue for now if nothing bad happens

**Comment by larry77** [comment]
Hi, 
The simulation completed successfully,  so you can close this. 
I'll let you know if I happen to have similar issues in the future.  

Lorenzo 
--

**Comment by tqchen** [comment]
great, thanks !
